sdk/src/types/api/core.rs

```diff
@@ -525,3 +526,84 @@ pub struct UtxoChangesResponse {
     pub created_outputs: Vec<OutputId>,
     pub consumed_outputs: Vec<OutputId>,
 }
+
+/// Contains the generic [`Output`] with associated [`OutputIdProof`].
+#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct OutputResponse {
+    pub output: Output,
+    pub output_id_proof: OutputIdProof,
+}
+
+/// The proof of the output identifier.
+#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct OutputIdProof {
+    pub slot: SlotIndex,
+    pub output_index: u16,
+    pub transaction_commitment: String,
+    pub output_commitment_proof: OutputCommitmentProof,
+}
+
+#[derive(Clone, Debug, Eq, PartialEq, Serialize)]
+#[serde(untagged)]
+pub enum OutputCommitmentProof {
+    HashableNode(HashableNode),
+    LeafHash(LeafHash),
+    ValueHash(ValueHash),
+}
+
+impl<'de> Deserialize<'de> for OutputCommitmentProof {
+    fn deserialize<D: serde::Deserializer<'de>>(d: D) -> Result<Self, D::Error> {
+        let value = Value::deserialize(d)?;
+        Ok(
+            match value
+                .get("type")
+                .and_then(Value::as_u64)
+                .ok_or_else(|| serde::de::Error::custom("invalid output commitment proof type"))?
+                as u8
+            {
+                0 => Self::HashableNode(
+                    serde_json::from_value::<HashableNode>(value)
+                        .map_err(|e| serde::de::Error::custom(format!("cannot deserialize hashable node: {e}")))?,
+                ),
+                1 => Self::LeafHash(
+                    serde_json::from_value::<LeafHash>(value)
+                        .map_err(|e| serde::de::Error::custom(format!("cannot deserialize leaf hash: {e}")))?,
+                ),
+                2 => Self::ValueHash(
+                    serde_json::from_value::<ValueHash>(value)
+                        .map_err(|e| serde::de::Error::custom(format!("cannot deserialize value hash: {e}")))?,
+                ),
+                _ => return Err(serde::de::Error::custom("invalid output commitment proof")),
+            },
+        )
+    }
+}
+
+/// The proof of the output commitment.
+#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
+pub struct HashableNode {
+    #[serde(rename = "type")]
+    pub kind: u8,
+    pub l: Box<OutputCommitmentProof>,
```

Comment:
r should be optional, right?

Reply:
Why optional? Would be just a LeafHash then, no?

Reply:
well one of these can be empty, iirc

Reply:
Don't see in which case that would be and also don't see anything optional here
https://github.com/iotaledger/iota.go/blob/2934050abf37c9ece6a35a9c5c96f87c648164b9/merklehasher/merkle_proof.go#L36

Reply:
I think Chloé is referring to the Rust POI impl in Chronicle, but that is a bit different. The proof for a single item (may it be a block or here an output) cannot just be a OutputCommitmentProof::ValueHash. It's still OutputCommitmentProof::LeafHash containing the OutputCommitmentProof::ValueHash. I don't think this can change due to how our merkle hasher works. So a value hash in that impl doesn't exist, and it contains either a node or a leaf, which is identical to one field being optional.

Reply:
still, is it not possible to have only one child?

Reply:
you can build a merkle tree over single element... by convention this is prefixed with `l` (left).

Reply:
I see!